Number_theory.hpp

```diff
@@ -0,0 +1,261 @@
+        return primes;
+    }
+
+    /*decomposes a number into powers of primes*/
+    inline auto decompose(std::integral auto value)
+    {
+        std::vector<std::pair<int, int>> primeFactors;
+
+        for (int i = 2; i <= value / i; ++i)
+        {
+            int count = 0;
+            while (value % i == 0)
+            {
+                ++count;
+                value /= i;
+            }
+            if (count > 0)
+            {
+                primeFactors.emplace_back(i, count);
+            }
+        }
+        if (value > 1)
+        {
+            primeFactors.emplace_back(value, 1);
+        }
+        return primeFactors;
+    }
+
+    /*calculates Euler�s Totient Function*/
+    inline auto euler_totient(std::integral auto value)
+    {
+        decltype(value) result = 0;
+        for (int i = 1; i <= value; i++) {
+            if (gcd(i, value) == 1) {
+                result++;
+            }
+        }
+        return result;
+    }
+
+    /*finds the largest power pow of prime such that prime^pow divides value*/
+    inline auto largest_power_of_prime_dividing_factorial(std::integral auto value, std::integral auto prime)
+    {
+        decltype(prime) pow = 0;
+        while (value % prime == 0)
+        {
+            pow++;
+            value /= prime;
+        }
+        return pow;
+    }
+
+    /*solves a system of linear equations as[i] * x = bs[i] mod mods[i].*/
+    template <std::integral I>
+    inline auto linear_congruence_solver(std::vector<I> as, std::vector<I> bs, std::vector<std::size_t> mods)
+    {
+        if (as.size() != bs.size() || bs.size() != mods.size())
+        {
+            throw std::invalid_argument("Input vectors must have the same size.");
+        }
+
+
+        for (size_t i = 0; i < as.size(); i++)
+        {
+
+            try
+            {
+                auto a_inverse = modular_inverse(as[i], mods[i]);
+                bs[i] = (bs[i] * a_inverse) % mods[i];
+            }
+            catch (const std::invalid_argument& e)
+            {
+                auto a = as[i];
+                auto mod = mods[i];
+                auto b = bs[i];
+
+
+                if (b % a == 0 && mod % a == 0)
+                {
+                    bs[i] = (b / a) % (mod / a);
+                    mods[i] = (mod / a);
+                }
+                else
+                {
+                    throw std::invalid_argument("Mods must be coprime");
+                }
+            }
+
+        }
+
+
+        if (!are_coprime(mods))
+        {
+            throw std::invalid_argument("Mods must be coprime");
+        }
+
+
+        std::size_t m = 1;
+        for (size_t i = 0; i < mods.size(); i++)
+        {
+            m *= mods[i];
+        }
+
+        auto result = 0;
+        for (size_t i = 0; i < as.size(); i++)
+        {
+            size_t mi = m / mods[i];
+            auto mi_inverse = modular_inverse(mi, mods[i]);
+            result += (bs[i] * mi_inverse * mi);
+        }

```

Comment:
this works only in the case where modulos are coprime - you should not assume this condition is true